### PR TITLE
final version of the Enumerable Methods

### PR DESCRIPTION
**Pull request by SpaYco**
##  Enumerable Methods
- Created #my_each
- Created #my_all?
- Created #my_any?
- Created #my_none?
- Created #my_count
- Created #my_map
- Created #my_inject
- No Linter Issues

### CODE REVIEW

**Comment by SpaYco** [comment]
I fixed the issues from the [previous PR](https://github.com/Spayco/Enumerables-Method/pull/1) 
I also changed the branch name so I had to create another PR, closed the previous one

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
+module Enumerable
+  def my_each
+    return Enumerator if block_given? == false
```
Comment (MauricioRobayo) [inline enumerables.rb]: 🤔 https://apidock.com/ruby/Object/to_enum

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Hi! Well done so far.

We need to double-check the following issues before continuing:

- [x] Make sure `my_each` returns an Enumerator if no block is given, take a look at the difference:

```ruby
>> [1,2].each
=> #<Enumerator: [1, 2]:each>
>> [1,2].my_each
=> Enumerator
```

- [x] Same for `my_each_with_index`.
- [x] Same for `my_select`.
- [x] Same for `my_map`
- [x] `my_all?` should return false when the block returns false or nil:

```ruby
>> [1,2].my_all? { |i| i != 2 }
=> true
>> [1,2].all? { |i| i != 2 }
=> false
>> [1,2].my_all?(String)
=> true
>> [1,2].all?(String)
=> false
```
- [x] `my_any?` should return true if the block ever returns a value other than false or nil:

```ruby
>> [1, 2].my_any? { |i| i == 2 }
=> false
>> [1, 2].any? { |i| i == 2 }
=> true
```

- [x] `my_any?` when no block or argument is given returns true the collection is not false or nil:

```ruby
>> [1].my_any?
=> false
>> [1].any?
=> true
```

- [x] `my_any?` there is an issue with Strings:

```ruby
%w[a b].my_any?('c')
```

- [x] `my_none?` should return true when the block returns true:

```ruby
>> [1,2].my_none? { |i| i == 2 }
=> true
>> [1,2].none? { |i| i == 2 }
=> false
```
- [x] `my_`none?` when no block or argument is given returns true only if none of the collection is true:

```ruby
>> [nil].my_none?
=> false
>> [nil].none?
=> true
````

- [x] `my_count` counts the number of elements yielding a true value if a block is given:

```ruby
>> [1, 2, 3 ,4].my_count { |i| i > 2 }
=> 4
>> [1, 2, 3 ,4].count { |i| i > 2 }
=> 2
```

- [x] `my_map` should accept a block, a proc, or both and when given both executes only the proc (check the [Odin project #12](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). The real map doesn't have this feature.

- [x] `my_inject` is throwing an error when given an initial value:

```ruby
>> [1, 2, 3].my_ inject(2, :+) # Error
>> array.my_inject(2) { |acc, nxt| acc + nxt } # Error
```

Status: __Request changes__.

**Comment by SpaYco** [comment]
fixed all the problems! (hopefully xD)

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Alright, it is getting there, but new issues raised:

* [x] `my_all?` returns true when none of the collection members are false or nil, for example:

```ruby
>> [1].my_all?
=> false
>> [1].all?
=> true
```

* [x] Same for `my_any?`:

```ruby
>> [1].my_any?
=> false
>> [1].any?
=> true
```

* [x] `my_any?` should return false if none of the collection matches the pattern, but for strings, it is throwing an error:

```ruby
>> %w[a b c].my_any?('d') # TypeError (type mismatch: String give)
```

* [x] `my_none?` when no block or argument is given returns true only if none of the collection members is true:

```ruby
>> [1].my_none?
=> true
>> [1].none?
=> false
```

* [x] `my_map` still not returning an Enumerator:

```ruby
>> [1].map
=> #<Enumerator: [1]:map>
>> [1].my_map # LocalJumpError (no block given (yield))
```
* [x] `my_map` still failing when both a block and a proc are given (should execute the proc). Check the [Odin project #12](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)). The real map doesn't have this feature:

```ruby
>> my_proc = Proc.new {|num| num > 10 }
>> [18, 22, 5, 6].my_map(my_proc) {|num| num < 10 } # true true false false
```

* [x] `my_inject` is now  throwing an error when applying a block:

```ruby
>> [1,1,1].inject { |prv, nxt| prv + nxt }
=> 3
>> [1,1,1].my_inject { |prv, nxt| prv + nxt } # TypeError (nil is not a symbol nor a string)
```

Keep strong, you're getting there!

Status: __Request changes__.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Spayco        
#### Great Job so far, but there are a few more things required. 

- [x] `my_map`  returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_map` should return `#<Enumerator>`
- [x] `my_inject`:
- Combines all elements of enum by applying a binary operation, specified by a block
`p (5..10).my_inject(4) { |prod, n| prod * n }` - should return 604800
- [x] `my_none?`:
- When no block or argument is given returns true only if none of the collection members is true.
`p [nil, false, true, []].my_none? == true`
`p [nil, false, true, []].none? == false`
- [x] `my_all?`:
-  when no block or argument is given returns true when none of the collection members are false or nil.
`p [1, true, 'hi', []].my_all? == false`
`p [1, true, 'hi', []].all? == true`
- [x] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil.
`p [1, true, 'hi', []].my_any? == false`
`p [1, true, 'hi', []].any? == true`
-  when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['dog','week'].my_any?('cat') == p ['dog','week'].any?('cat') should return true`













### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Awesome, you got it, it is almost there. Please double-check:

* [x] `my_map` returns a new array with the results of running block once for every element in enum:

```ruby
>> [1, 2].map { |num| num > 1 }
=> [false, true]
>> [1, 2].my_map { |num| num > 1 }
=> #<Enumerator: [1, 2]:each>
```

* [x] `my_inject` is throwing an error when a collection of strings given:

```ruby
>> %w[hello, world].inject { |prv, nxt| prv + nxt }
=> "hello,world"
>> %w[hello, world].my_inject { |prv, nxt| prv + nxt } # TypeError (String can't be coerced into Integer)
```

Status: __Request changes__.

**Comment by SpaYco** [comment]
should be good to go

**Review by MauricioRobayo — APPROVED** [review]
You got it, well done!

The project looks great and all requirements are met, please move to bigger challenges.

Status: __Approved__.

**Pull request merged**